File: typescript/src/activities.ts
import { ApplicationFailure, Context, log, sleep } from '@temporalio/activity';
import type { BookTripInput } from './shared';

export async function bookFlight(input: BookTripInput): Promise<string> {
  log.info(`Booking flight: ${input.flightId}`);

  await sleep(1000);

  if (input.flightId.toLowerCase().includes('flaky')) {
    // a transient error, which will be retried
    if (Context.current().info.attempt < 6) {
      throw ApplicationFailure.create({
        message: 'Flight booking service is currently unavailable',
      });
    }
  }

  return `Booked flight: ${input.flightId}`;
}

export async function bookHotel(input: BookTripInput): Promise<string> {
  log.info(`Booking hotel: ${input.hotelId}`);

  await sleep(1000);

  if (input.hotelId.toLowerCase().includes('buggy')) {
    // a simulated bug
    const error = true;
    if (error) {
      throw new Error('Error due to bug in code');
    }
  }

  return `Booked hotel: ${input.hotelId}`;
}

export async function bookCar(input: BookTripInput): Promise<string> {
  log.info(`Booking car ${input.carId}`);

  await sleep(1000);

  if (input.carId.toLowerCase().includes('invalid')) {
    // a business error, which cannot be retried
    throw ApplicationFailure.create({
      nonRetryable: true,
      message: `Car ${input.carId} is invalid`,
      type: 'InvalidCar',
    });
  }

  return `Booked car: ${input.carId}`;
}

export async function notifyUser(input: BookTripInput): Promise<string> {
  log.info(`Notifying user: ${input.userId}`);

  await sleep(1000);

  return `Notified user: ${input.userId}`;
}

export async function undoBookFlight(input: BookTripInput): Promise<string> {
  log.info(`Undo flight booking: ${input.flightId}`);

  await sleep(1000);

  return `Unbooked flight: ${input.flightId}`;
}

export async function undoBookHotel(input: BookTripInput): Promise<string> {
  log.info(`Undo hotel booking: ${input.hotelId}`);

  await sleep(1000);

  return `Unbooked hotel: ${input.hotelId}`;
}

export async function undoBookCar(input: BookTripInput): Promise<string> {
  log.info(`Undo car booking: ${input.carId}`);

  await sleep(1000);

  return `Unbooked car: ${input.carId}`;
}

File: ruby/src/activities.rb
# frozen_string_literal: true

require_relative 'shared'
require 'temporalio/activity'

module TripBooking
  module Activities
    class BookFlight < Temporalio::Activity::Definition
      def execute(input)
        Temporalio::Activity::Context.current.logger.info("Booking flight: #{input.flightId}")
        sleep(1)

        if input.flightId.to_s.downcase.include?('flaky')
          # a transient error, which will be retried
          attempt = Temporalio::Activity::Context.current.info.attempt
          raise 'Flight booking service is currently unavailable' if attempt < 6
        end

        "Booked flight: #{input.flightId}"
      end
    end

    class BookHotel < Temporalio::Activity::Definition
      def execute(input)
        Temporalio::Activity::Context.current.logger.info("Booking hotel: #{input.hotelId}")
        sleep(1)

        if input.hotelId.to_s.downcase.include?('buggy')
          # a simulated bug
          error = true
          raise 'Error due to bug in code' if error
        end

        "Booked hotel: #{input.hotelId}"
      end
    end

    class BookCar < Temporalio::Activity::Definition
      def execute(input)
        Temporalio::Activity::Context.current.logger.info("Booking car: #{input.carId}")
        sleep(1)

        if input.carId.to_s.downcase.include?('invalid')
          # a business error, which cannot be retried
          raise Temporalio::Error::ApplicationError.new(
            "Car #{input.carId} is invalid",
            non_retryable: true
          )
        end

        "Booked car: #{input.carId}"
      end
    end

    class NotifyUser < Temporalio::Activity::Definition
      def execute(input)
        Temporalio::Activity::Context.current.logger.info("Notifying user: #{input.userId}")
        sleep(1)
        "Notified user: #{input.userId}"
      end
    end

    class UndoBookFlight < Temporalio::Activity::Definition
      def execute(input)
        Temporalio::Activity::Context.current.logger.info("Undo flight booking: #{input.flightId}")
        sleep(1)
        "Unbooked flight: #{input.flightId}"
      end
    end

    class UndoBookHotel < Temporalio::Activity::Definition
      def execute(input)
        Temporalio::Activity::Context.current.logger.info("Undo hotel booking: #{input.hotelId}")
        sleep(1)
        "Unbooked hotel: #{input.hotelId}"
      end
    end

    class UndoBookCar < Temporalio::Activity::Definition
      def execute(input)
        Temporalio::Activity::Context.current.logger.info("Undo car booking: #{input.carId}")
        sleep(1)
        "Unbooked car: #{input.carId}"
      end
    end
  end
end
